.NET Getting started - based on roll dice service and Automatic Instrumentation (#3167)
open-telemetry/opentelemetry.io: content/en/docs/instrumentation/net/getting-started.md

You will learn how you can instrument a simple .NET application automatically,
in such a way that [traces][], [metrics][] and [logs][] are emitted to the
console.

## Prerequisites

Ensure that you have the following installed locally:

- [.NET SDK](https://dotnet.microsoft.com/download/dotnet) 6+

## Example Application

The following example uses a basic
[Minimal API with ASP.NET Core](https://learn.microsoft.com/aspnet/core/tutorials/min-web-api)
application. If you are not using ASP.NET Core, that's OK — you can still use
OpenTelemetry .NET Automatic Instrumentation.

For more elaborate examples, see
[examples](/docs/instrumentation/net/examples/).

### Create and launch an HTTP Server

To begin, set up an environment in a new directory called `dotnet-simple`.
Within that directory, execute following command:

```sh
dotnet new web
```

In the same directory, replace the content of `Program.cs` with the following
code:

```csharp
using System.Globalization;

var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

var logger = app.Logger;
var random = new Random();

app.MapGet("/rolldice/{player?}", HandleRollDice);

app.Run();

> return;

Reviewer: Mmmm, I don't really like this. I'd much prefer either:

1. The handler moves up pass `app.Run()`
2. The handler defined inline
3. The handler in another file

I think it's just a little too subtle to have the `return` statement there.

Reviewer: What do you think?

Author: I have started with your option 1:
```csharp
using System.Globalization;

var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

var logger = app.Logger;
var random = new Random();

int RollDice()
{
    return random.Next(1, 7);
}

string HandleRollDice(string? player)
{
    var result = RollDice();

    if (string.IsNullOrEmpty(player))
    {
        logger.LogInformation("Anonymous player is rolling the dice: {result}", result);
    }
    else
    {
        logger.LogInformation("{player} is rolling the dice: {result}", player, result);
    }

    return result.ToString(CultureInfo.InvariantCulture);
}

app.MapGet("/rolldice/{player?}", HandleRollDice);

app.Run();
```

Then checked all suggestion in IDE (VS2023+R# with default settings) and applied all reformat suggestions.
It leads to the current state (it was mainly fixing `Put local function '{name}' after 'return' statement`).

I am fine with both solutions. If you want, I can add commit with this version. Code is simple enough in both cases.

I would avoid creating separate file or inlining methods. I think one file with some logical separation is good compromise for this kind of examples.

Reviewer: Mmmm, I should ping a friend on the team to complain to them about that refactoring :)

Yeah I'd prefer Option 1 if possible -- I think it's fine for people to use R# refactors on their own, but I do think it makes the code more subtle to understand, and we want things to be as straightforward as possible on the docs site (I also acknowledge that "straightforward" can be subjective!)

int RollDice()
{
    return random.Next(1, 7);
}

string HandleRollDice(string? player)
{
    var result = RollDice();

    if (string.IsNullOrEmpty(player))
    {
        logger.LogInformation("Anonymous player is rolling the dice: {result}", result);
    }
    else
    {
        logger.LogInformation("{player} is rolling the dice: {result}", player, result);
    }

    return result.ToString(CultureInfo.InvariantCulture);
}
```

In the `Properties` subdirectory, replace the content of `launchSettings.json`
with the following:

```json
{
  "$schema": "http://json.schemastore.org/launchsettings.json",
  "profiles": {
    "http": {
      "commandName": "Project",
      "dotnetRunMessages": true,
      "launchBrowser": true,
      "applicationUrl": "http://localhost:8080",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    }
  }
}
```

Build and run the application with the following command, then open
<http://localhost:8080/rolldice> in your web browser to ensure it is working.

```sh
dotnet build
dotnet run
```

> ## Instrumentation

Reviewer: For this section, I'd also like to see a Windows tab for the install link and command. Similarly, the env vars should have an appropriate Windows section. I think a ton of .NET developers will try stuff out on a Windows machine but then deploy to Linux, and the current link won't work unless they're on WSL.

Author: Good point. To be honest, I was using github shell on Windows to write and tests this document so I have missed that PowerShell also covered. It should work on all systems if you have linux-like shell.

Author: Fixed. PowerShell scripts tested on newly created Windows 11 Azure. The only manual changes - installation .NET SDK 6.

---
title: Getting Started
description: Get telemetry for your app in less than 5 minutes!
cSpell:ignore: ASPNETCORE rolldice
weight: 10
dotnet-auto-install-sh-URL: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/releases/latest/download/otel-dotnet-auto-install.sh

> dotnet-auto-install-ps-URL: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/releases/latest/download/OpenTelemetry.DotNet.Auto.psm1

Author: Fixed 6d132a9c28c0e47f2284a76868b9a0d0f22f1422

This page will show you how to get started with OpenTelemetry in .NET.

You will learn how you can instrument a simple .NET application automatically,
in such a way that [traces][], [metrics][] and [logs][] are emitted to the
console.

## Prerequisites

Ensure that you have the following installed locally:

- [.NET SDK](https://dotnet.microsoft.com/download/dotnet) 6+

## Example Application

The following example uses a basic
[Minimal API with ASP.NET Core](https://learn.microsoft.com/aspnet/core/tutorials/min-web-api)
application. If you are not using ASP.NET Core, that's OK — you can still use
OpenTelemetry .NET Automatic Instrumentation.

For more elaborate examples, see
[examples](/docs/instrumentation/net/examples/).

### Create and launch an HTTP Server

To begin, set up an environment in a new directory called `dotnet-simple`.
Within that directory, execute following command:

```sh
dotnet new web
```

In the same directory, replace the content of `Program.cs` with the following
code:

```csharp
using System.Globalization;

var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

var logger = app.Logger;
var random = new Random();

int RollDice()
{

>     return random.Next(1, 7);

Reviewer 2: nit: Use `Random.Shared` instead of a dedicated `Random` instance.

}

app.MapGet("/rolldice/{player?}", HandleRollDice);

app.Run();
```

In the `Properties` subdirectory, replace the content of `launchSettings.json`
with the following:

```json
{
  "$schema": "http://json.schemastore.org/launchsettings.json",
  "profiles": {
    "http": {
      "commandName": "Project",
      "dotnetRunMessages": true,
      "launchBrowser": true,
      "applicationUrl": "http://localhost:8080",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    }
  }
}
```

Build and run the application with the following command, then open
<http://localhost:8080/rolldice> in your web browser to ensure it is working.

```sh
dotnet build
dotnet run
```

## Instrumentation

Next, you'll use a [OpenTelemetry .NET Automatic Instrumentation](../automatic)
to instrument the application at launch time. While you can [configure .NET
Automatic Instrumentation][] in a number of ways, the steps below use Unix-shell
or PowerShell scripts.

> **Note**: PowerShell commands require elevated (administrator) privileges.

1. Download installation scripts from [Releases][] of the
   `opentelemetry-dotnet-instrumentation` repository:

>    {{< tabpane text=true >}} {{% tab Unix-shell %}}

Reviewer 2: What are the `tabpane` and `tab` added for? They are plainly rendered as text in the rich diff view.

Reviewer: It's docs syntax to support tabbed content, see here: https://deploy-preview-3167--opentelemetry.netlify.app/docs/instrumentation/net/getting-started/#instrumentation

   ```sh
   curl -L -O {{% _param dotnet-auto-install-sh-URL %}}
   ```

   {{% /tab %}} {{% tab PowerShell - Windows %}}

   ```powershell
   $module_url = "{{% _param dotnet-auto-install-ps-URL %}}"
   $download_path = Join-Path $env:temp "OpenTelemetry.DotNet.Auto.psm1"
   Invoke-WebRequest -Uri $module_url -OutFile $download_path -UseBasicParsing
   ```

   {{% /tab %}} {{< /tabpane >}}

2. Execute following script to download automatic instrumentation for your
   development environment:

   {{< tabpane text=true >}} {{% tab Unix-shell %}}

   ```sh
   ./otel-dotnet-auto-install.sh
   ```

   {{% /tab %}} {{% tab PowerShell - Windows %}}

   ```powershell
   Import-Module $download_path
   Install-OpenTelemetryCore
   ```

   {{% /tab %}} {{< /tabpane >}}

3. Set and export variables that specify a [console exporter][], then execute

>    script configuring other necessary environmental variables using a notation

Reviewer 2: ```suggestion
   script configuring other necessary environment variables using a notation
```

Author: Fixed 6d132a9c28c0e47f2284a76868b9a0d0f22f1422

     OTEL_DOTNET_AUTO_TRACES_CONSOLE_EXPORTER_ENABLED=true \
     OTEL_DOTNET_AUTO_METRICS_CONSOLE_EXPORTER_ENABLED=true \
     OTEL_DOTNET_AUTO_LOGS_CONSOLE_EXPORTER_ENABLED=true
     OTEL_SERVICE_NAME=RollDiceService
   . $HOME/.otel-dotnet-auto/instrument.sh
   ```

   {{% /tab %}} {{% tab PowerShell - Windows %}}

   ```powershell
   $env:OTEL_TRACES_EXPORTER="none"
   $env:OTEL_METRICS_EXPORTER="none"
   $env:OTEL_LOGS_EXPORTER="none"
   $env:OTEL_DOTNET_AUTO_TRACES_CONSOLE_EXPORTER_ENABLED="true"
   $env:OTEL_DOTNET_AUTO_METRICS_CONSOLE_EXPORTER_ENABLED="true"
   $env:OTEL_DOTNET_AUTO_LOGS_CONSOLE_EXPORTER_ENABLED="true"
   Register-OpenTelemetryForCurrentSession -OTelServiceName "RollDiceService"
   ```

   {{% /tab %}} {{< /tabpane >}}

4. Run your **application** once again:

   ```sh
   dotnet run
   ```

   Note the output from the `dotnet run`.

5. From _another_ terminal, send a request using `curl`:

   ```sh
   curl localhost:8080/rolldice
   ```

6. After about 30 sec, stop the server process.

At this point, you should see trace and log output from the server and client
that looks something like this (output is line-wrapped for readability):

<details>

> <summary>Traces and logs</summary>

Reviewer 2: ```suggestion
<summary>Traces and Logs</summary>
```

Author: Fixed 6d132a9c28c0e47f2284a76868b9a0d0f22f1422

   %}}) from [Releases][] of the `opentelemetry-dotnet-instrumentation` repository:

   ```sh
   curl -L -O {{% _param dotnet-auto-install-URL %}}
   ```

2. Execute `otel-dotnet-auto-install.sh` script to download automatic
   instrumentation for your development environment:

   ```sh
   ./otel-dotnet-auto-install.sh
   ```

3. Set and export variables that specify a [console exporter][], then execute
   shell script configuring other necessary environmental variables using a
   notation suitable for your shell/terminal environment &mdash; we illustrate a
   notation for bash-like shells:

   ```sh
   export OTEL_TRACES_EXPORTER=none \
     OTEL_METRICS_EXPORTER=none \
     OTEL_LOGS_EXPORTER=none \
     OTEL_DOTNET_AUTO_TRACES_CONSOLE_EXPORTER_ENABLED=true \
     OTEL_DOTNET_AUTO_METRICS_CONSOLE_EXPORTER_ENABLED=true \
     OTEL_DOTNET_AUTO_LOGS_CONSOLE_EXPORTER_ENABLED=true
   . $HOME/.otel-dotnet-auto/instrument.sh
   ```

4. Run your **application** once again:

   ```sh
   dotnet run
   ```

   Note the output from the `dotnet run`.

5. From _another_ terminal, send a request using `curl`:

   ```sh
   curl localhost:8080/rolldice
   ```

> 6. Stop the server process.

Reviewer 3: I agree with @cijothomas (https://github.com/open-telemetry/opentelemetry.io/pull/3167#discussion_r1293819921), consider:

```suggestion
6. After about 30 sec, stop the server process.
```

> At step 6, you should have seen trace and log output from the server and client
> that looks something like this (output is line-wrapped for readability):

Reviewer 3: ```suggestion
At this point, you should see trace and log output from the server and client
that looks something like this (output is line-wrapped for readability):
```

> ```log

Reviewer 3: Nice to have: this sample output is a bit long, consider wrapping it in

```html
<details>
<summary>View output</summary>
...
</details>
```

---
title: Getting Started
description: Get telemetry for your app in less than 5 minutes!
cSpell:ignore: ASPNETCORE rolldice
weight: 10
---

This page will show you how to get started with OpenTelemetry in .NET.

You will learn how you can instrument a simple .NET application automatically,
in such a way that [traces][], [metrics][] and [logs][] are emitted to the
console.

## Prerequisites

Ensure that you have the following installed locally:

- [.NET SDK](https://dotnet.microsoft.com/download/dotnet) 6+

## Example Application

The following example uses a basic
[Minimal API with ASP.NET Core](https://learn.microsoft.com/aspnet/core/tutorials/min-web-api)
application. If you are not using ASP.NET Core, that's OK — you can still use
OpenTelemetry .NET Automatic Instrumentation.

For more elaborate examples, see
[examples](/docs/instrumentation/net/examples/).

> ### Dependencies

Reviewer 4: what does the heading "Dependencies" mean?

Author: Leftover from java structure.
Merged with Create and luaunch an HTTP server 323f3cd22e67bbe176e7fc41242d075236bf37c6

int RollDice()
{
    return random.Next(1, 7);
}

string HandleRollDice(string? player)
{
    var result = RollDice();

    if (string.IsNullOrEmpty(player))
    {
        logger.LogInformation("Anonymous player is rolling the dice: {result}", result);
    }
    else
    {
        logger.LogInformation("{player} is rolling the dice: {result}", player, result);
    }

    return result.ToString(CultureInfo.InvariantCulture);
}

```

In the `Properties` subdirectory, modify a file called `launchSettings.json` and
replace existing content by the following code:

```json
{
  "$schema": "http://json.schemastore.org/launchsettings.json",
  "profiles": {
    "http": {
      "commandName": "Project",
      "dotnetRunMessages": true,
      "launchBrowser": true,
      "applicationUrl": "http://localhost:8080",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    }
  }
}
```

Build and run the application with the following command, then open
<http://localhost:8080/rolldice> in your web browser to ensure it is working.

```sh
dotnet build
dotnet run
```

## Instrumentation

Next, you'll use a [OpenTelemetry .NET Automatic Instrumentation](../automatic)

> to instrument the application at launch time. While you can [configure .NET

Reviewer 4: would be nice to describe what gets automatically instrumented here.. (the asp.net core server itself, any httpclient calls, etc.. so users know why they are seeing (or not seeing) the telemetry at the end.)

Author: I think that it is to early for details. It can go to ../automatic somehow.

I expect form getting started is to present that traces, metrics, and logs can be handled by OTel. Details are not so important.

Reviewer: Agreed, we can elaborate more on what's instrumented later. I think it will also be pretty obvious what's instrumented when people run a sample app like in the docs here.

Automatic Instrumentation][] in a number of ways, the steps below uses shell
scripts.

1. Download [otel-dotnet-auto-install.sh][] from [Releases][] of the
   `opentelemetry-dotnet-instrumentation/releases` repository. The file contains
   the script which detect system and download appropriate instrumentation
   version:

   ```sh

>    curl -L -O https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/releases/latest/download/otel-dotnet-auto-install.sh

Reviewer 3: ```suggestion
   curl -L -O {{% _param dotnet-auto-install-URL %}}
```

   ```

2. Execute `otel-dotnet-auto-install.sh` script to download automatic
   instrumentation:

   ```sh
   ./otel-dotnet-auto-install.sh
   ```

3. Set and export variables that specify a [console exporter][], then execute
   shell script configuring other necessary environmental variables using a
   notation suitable for your shell/terminal environment &mdash; we illustrate a
   notation for bash-like shells:

   ```sh
   export OTEL_TRACES_EXPORTER=none \
     OTEL_METRICS_EXPORTER=none \
     OTEL_LOGS_EXPORTER=none \
     OTEL_DOTNET_AUTO_TRACES_CONSOLE_EXPORTER_ENABLED=true \
     OTEL_DOTNET_AUTO_METRICS_CONSOLE_EXPORTER_ENABLED=true \
     OTEL_DOTNET_AUTO_LOGS_CONSOLE_EXPORTER_ENABLED=true
   . $HOME/.otel-dotnet-auto/instrument.sh
   ```

4. Run your **application** once again:

   ```sh
   dotnet run
   ```

   Note the output from the `dotnet run`.

5. From _another_ terminal, send a request using `curl`:

   ```sh
   curl localhost:8080/rolldice
   ```

6. Stop the server process.

> At step 6, you should have seen trace & log output from the server and client
> that looks something like this (trace output is line-wrapped for convenience):

Reviewer 3: ```suggestion
At step 6, you should have seen trace and log output from the server and client
that looks something like this (output is line-wrapped for readability):
```

Resource associated with LogRecord:
service.name: simple-dotnet
telemetry.auto.version: 0.7.0
telemetry.sdk.name: opentelemetry
telemetry.sdk.language: dotnet
telemetry.sdk.version: 1.4.0.802

info: simple-dotnet[0]
      Anonymous player is rolling the dice: 1
Activity.TraceId:            3961d22b5f90bf7662ad4933318743fe
Activity.SpanId:             93d5fcea422ff0ac
Activity.TraceFlags:         Recorded
Activity.ActivitySourceName: OpenTelemetry.Instrumentation.AspNetCore
Activity.DisplayName:        /rolldice
Activity.Kind:               Server
Activity.StartTime:          2023-08-14T06:44:53.9278162Z
Activity.Duration:           00:00:00.0049754
Activity.Tags:
    net.host.name: localhost
    net.host.port: 8080
    http.method: GET
    http.scheme: http
    http.target: /rolldice
    http.url: http://localhost:8080/rolldice
    http.flavor: 1.1
    http.user_agent: curl/8.0.1
    http.status_code: 200
Resource associated with Activity:
    service.name: simple-dotnet
    telemetry.auto.version: 0.7.0
    telemetry.sdk.name: opentelemetry
    telemetry.sdk.language: dotnet
    telemetry.sdk.version: 1.4.0.802
```

> At step 6, when stopping the server, you should see an output of all the metrics

Reviewer 4: nit: the metric console exporter should be exporter every few seconds, so maybe put a note saying app should be run for 20-30 seconds, so as to see metrics as well..(that would be better for all signals, as OTLP exporter for trace/log also does the batching)

Author: Part of c8693f4f3d7da9affbc4d69fc5bc9abedaf92f21

> At step 6, when stopping the server, you should see an output of all the metrics
> collected (metrics output is shortened for convenience):

Reviewer 3: ```suggestion
Also when stopping the server, you should see an output of all the metrics
collected (sample excerpt shown):
```

```log
Export process.runtime.dotnet.gc.collections.count, Number of garbage collections that have occurred since process start., Meter: OpenTelemetry.Instrumentation.Runtime/1.1.0.2
(2023-08-14T06:12:05.8500776Z, 2023-08-14T06:12:23.7750288Z] generation: gen2 LongSum
Value: 2
(2023-08-14T06:12:05.8500776Z, 2023-08-14T06:12:23.7750288Z] generation: gen1 LongSum
Value: 2
(2023-08-14T06:12:05.8500776Z, 2023-08-14T06:12:23.7750288Z] generation: gen0 LongSum
Value: 6

...

Export http.client.duration, Measures the duration of outbound HTTP requests., Unit: ms, Meter: OpenTelemetry.Instrumentation.Http/1.0.0.0

> (2023-08-14T06:12:06.2661140Z, 2023-08-14T06:12:23.7750388Z] http.flavor: 1.1 http.method: POST http.scheme: https http.status_code: 200 net.peer.name: dc.services.visualstudio.com Histogram

Reviewer 4: interesting metric! it looks like the app is connecting to application insights (dc.services url), and that got collected as http client metric.

Reviewer 4: the app itself is not making any http-client calls, so this could be confusing to see in the output!

Reviewer: Yeah agreed. I'm probably okay with leaving it, but ideally this wouldn't connect to any external service so that people who just run locally can easily compare outputs to their own.

Reviewer 5: This looks like telemetry from dotnet CLI, there is a `dotnet run` step. It may go away if `DOTNET_CLI_TELEMETRY_OPTOUT` is set to `true`.

Value: Sum: 1330.4766000000002 Count: 5 Min: 50.0333 Max: 465.7936
(-Infinity,0]:0
(0,5]:0
(5,10]:0
(10,25]:0
(25,50]:0
(50,75]:2
(75,100]:0
(100,250]:0
(250,500]:3
(500,750]:0
(750,1000]:0
(1000,2500]:0
(2500,5000]:0
(5000,7500]:0
(7500,10000]:0
(10000,+Infinity]:0
```

## What next?

For more:

- Run this example with another [exporter][] for telemetry data.
- Try [automatic instrumentation](../automatic/) on one of your own apps.
- Learn about [manual instrumentation][] and try out more

>   [examples](/docs/instrumentation/java/examples/).

Reviewer 3: Copy-paste error:

```suggestion
  [examples](/docs/instrumentation/net/examples/).
```

- Take a look at the [OpenTelemetry Demo](/docs/demo/), which includes .NET
  based [Cart Service](/docs/demo/services/cart/).

[traces]: /docs/concepts/signals/traces/
[metrics]: /docs/concepts/signals/metrics/
[logs]: /docs/concepts/signals/logs/
[configure .NET Automatic Instrumentation]: ../automatic
[console exporter]:
  https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/docs/config.md#internal-logs
[exporter]:
  https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/docs/config.md#exporters
[manual instrumentation]: ../manual

> [otel-dotnet-auto-install.sh]:
>   https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/releases/latest/download/otel-dotnet-auto-install.sh

Reviewer 3: ```suggestion
```
